src/StackExchange.Redis/PhysicalBridge.cs

```diff
@@ -750,45 +735,66 @@ internal WriteResult WriteMessageTakingWriteLockSync(PhysicalConnection physical
+            // Which StartBacklogProcessor will check.
+            StartBacklogProcessor();
             return true;
         }
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private void StartBacklogProcessor()
         {
+            if (Interlocked.CompareExchange(ref _backlogProcessorIsRunning, 1, 0) == 0)
+            {
+                
 #if DEBUG
-            _backlogProcessorRequestedTime = Environment.TickCount;
+                _backlogProcessorRequestedTime = Environment.TickCount;
 #endif
-            Task.Run(ProcessBacklogAsync);
+                Task.Run(ProcessBacklogAsync);
+            }
         }
 #if DEBUG
         private volatile int _backlogProcessorRequestedTime;
 #endif
 
         private void CheckBacklogForTimeouts() // check the head of the backlog queue, consuming anything that looks dead
         {
-            lock (_backlog)
+            var now = Environment.TickCount;
+            var timeout = TimeoutMilliseconds;
+
+            // Because peeking at the backlog, checking message and then dequeueing, is not thread-safe, we do have to use
+            // a lock here, for mutual exclusion of backlog DEQUEUERS. Unfortunately.
+            // But we can optimize to reduce contention by only locking if we see something that looks timed out.
+            Message message;
+            while (_backlog.TryPeek(out message))
             {
-                var now = Environment.TickCount;
-                var timeout = TimeoutMilliseconds;
-                while (_backlog.Count != 0)
+                if (message.IsInternalCall) break; // don't stomp these (not that they should have the async timeout flag, but...)
+                if (!message.HasAsyncTimedOut(now, timeout, out var _)) break; // not a timeout - we can stop looking
+                lock (_backlog)
                 {
-                    var message = _backlog.Peek();
-                    if (message.IsInternalCall) break; // don't stomp these (not that they should have the async timeout flag, but...)
+                    // peek again since we didn't have lock before...
+                    // and rerun the exact same checks as above
+                    if (!_backlog.TryPeek(out message)) break;
+                    if (message.IsInternalCall) break;
+                    if (!message.HasAsyncTimedOut(now, timeout, out var _)) break;
 
-                    if (!message.HasAsyncTimedOut(now, timeout, out var _)) break; // not a timeout - we can stop looking
-                    _backlog.Dequeue(); // consume it for real
+                    if (!_backlog.TryDequeue(out var message2) || (message != message2)) // consume it for real
+                    {
+                        throw new Exception("Thread safety bug detected! A queue message disappeared while we had the backlog lock");
+                    }
 
+                    // TODO: PERF, consider doing this outside the lock, or just taking the lock once and looping
```

Comment:
Noting for discussion on Tuesday!

Reply:
Whether perf matter here was a bit of a puzzle for me to be honest. I'm not sure if perf is the main concern for CheckBacklogForTimeouts, actually, unless there *are* timeouts, and then maybe its goodness that it holds the lock and processes as many timeouts as possible without letting anybody else enqueue them. 

However, now that I look at it again, I wonder how safe calling SetExceptionAndComplete while holding the lock is. Do continuations get to run on this thread? Could it ever cause deadlocks?